app/controllers/incomeSources/add/BusinessAddedObligationsController.scala

```diff
@@ -51,40 +54,70 @@ class BusinessAddedObligationsController @Inject()(authenticate: AuthenticationP
                                                    dateService: DateServiceInterface)
   extends ClientConfirmedController with I18nSupport with FeatureSwitching with IncomeSourcesUtils {
 
-  private def getBusinessNameAndStartDate(incomeSourceId: String)(implicit user: MtdItUser[_]): Option[(String, LocalDate)] = {
-    user.incomeSources.businesses
-      .find(_.incomeSourceId.equals(incomeSourceId))
-      .flatMap { addedBusiness =>
-        for {
-          businessName <- addedBusiness.tradingName
-          startDate <- addedBusiness.tradingStartDate
-        } yield (businessName, startDate)
-      }
+  def getBusinessNameAndStartDate(incomeSource: Either[Throwable, Option[String]])(implicit user: MtdItUser[_]): Either[Throwable, (String, LocalDate)] = {
+    incomeSource match {
+      case Right(Some(incomeSourceId)) =>
+        val addedBusinessDetails = for {
+          addedBusiness <- user.incomeSources.businesses.find(_.incomeSourceId.equals(incomeSourceId))
+          tradeName <- addedBusiness.tradingName
+          tradeStartDate <- addedBusiness.tradingStartDate
+        } yield (tradeName, tradeStartDate)
+
+        val res = addedBusinessDetails match {
+          case Some(businessDetails) => Right((businessDetails._1, businessDetails._2))
+          case None => Left(throw new Exception(s"Failed to get addedBusinessDetails"))
+        }
+        res
+      case Right(None) => Left(throw new Exception(s"Failed to get IncomeSource from session"))
```

Comment:
* you are attempting to throwing exception within block which expected to return error, this would cause opposite behaviour than expected;
* you can use next version for this method instead:

```
  def getBusinessNameAndStartDate(incomeSource: Either[Throwable, Option[String]])(implicit user: MtdItUser[_]): Either[Throwable, (String, LocalDate)] = {
    incomeSource match {
      case Right(Some(incomeSourceId)) => {
        for {
          addedBusiness <- user.incomeSources.businesses.find(_.incomeSourceId.equals(incomeSourceId))
          tradeName <- addedBusiness.tradingName
          tradeStartDate <- addedBusiness.tradingStartDate
        } yield Right((tradeName, tradeStartDate))
      }.getOrElse(Left(throw new Exception(s"Failed to get addedBusinessDetails: ${incomeSourceId}")))
      case Right(None) =>
        Left(new Error(s"Failed to extract incomeSource from session"))
      case Left(ex) =>
        Left(ex)
    }
  }
```